engine/menu/field_debug_menu.asm

```diff
@@ -0,0 +1,518 @@
+ITEM_TEST        EQU 23 ; unused
+ACCESS_PC        EQU 24
+ACCESS_MART      EQU 25
+SET_TELEPORT     EQU 26
+VRAM_VIEWER      EQU 27
+POKE_GEAR        EQU 28
+
+FieldDebugMenuPages:
+
+FieldDebugMenuPage1: ; 3F:40CE
+	db 7
+	db NEXT_PAGE
+	db WARP
+	db CHAR_VIEWER
+	db CHANGE_TRANSPORT
+	db TOOL_GEAR
+	db ACCESS_PC
+	db CLOSE
+	db -1
+
+FieldDebugMenuPage2: ; 3F:40D7
+	db 7
+	db NEXT_PAGE
+	db ACCESS_MART
+	db RECOVER_PKMN
+	db POKE_GEAR
+	db GAME_SELECT
+	db MAP_VIEWER
+	db CLOSE
+	db -1
+
+FieldDebugMenuPage3: ; 3F:40E0
+	db 7
+	db NEXT_PAGE
+	db EVENT_FLAGS
+	db VRAM_VIEWER
+	db SET_TELEPORT
+	db FOLLOW_NPC
+	db SHOW_MAP
+	db CLOSE
+	db -1
+
+SFX_MENU EQU 3 ; Menu opening sound
+
+FieldDebugMenu:: ; 3F:40E9
+	call RefreshScreen
+	ld de, SFX_MENU
+	call PlaySFX
+	ld hl, FieldDebugMenuHeader
+	call LoadMenuHeader
+	ld a, $00
+	ld [wFieldDebugMenuPage], a
+
+.Reopen: ; 3F:40FD
+	call UpdateTimePals
+	call UpdateSprites
+	ld a, [wFieldDebugMenuCursorBuffer]
+	ld [wMenuCursorBuffer], a
+	call OpenMenu
+	jr c, .Exit
+	ld a, [wMenuCursorBuffer]
+	ld [wFieldDebugMenuCursorBuffer], a
+	call PlaceHollowCursor
+	ld a, [wMenuJoypad]
+	cp A_BUTTON
+	jr z, .RunFunction
+	call ChangeFieldDebugMenuPage
+	jr .RunReturn
+
+.RunFunction: ; 3F:4123
+	ld a, [wMenuSelection]
+	ld hl, FieldDebugJumpTable
+	call CallJumptable
+
+; Field Debug menu items may have different return functions.
+.RunReturn: ; 3F:412C
+	ld hl, FieldDebugMenuReturns
+	jp CallJumptable
+
+.FieldDebugMenuReturns: ; 3F:4132
+	dw .Reopen
+	dw .Exit
+	dw .Close
+	dw .Clean
+	dw .Return
+
+.Exit: ; 3F:413C
+	call GetJoypad
+	ldh a, [hJoyDown]
+	bit 0, a
+	jr nz, .Exit
+	call LoadFontExtra
+
+.Close: ; 3F:4148
+	call CloseWindow
+
+.Clean: ; 3F:414B
+	push af
+	call Function1fea
+	pop af
+	ret
+
+.Return: ; 3F:4151
+	call ExitMenu
+	ld a, HMENURETURN_ASM
+	ld [hStartmenuCloseAndSelectHookEnable], a
+	jr .Clean
+
+FieldDebug_CloseFieldDebugMenu: ; 3F:415A
+	ld a, $02
+	ret
+
+ChangeFieldDebugMenuPage: ; 3F:415D
+	ld a, [wMenuJoypad]
+	cp D_LEFT
+	jr z, .prev_page
+.next_page
+	ld a, [wFieldDebugMenuPage]
+	inc a
+	cp $03
+	jr nz, .load_next_page
+.set_page_0
+	xor a
+.load_next_page
+	ld [wFieldDebugMenuPage], a
+	jr PlayMenuOpeningSound
+.prev_page
+	ld a, [wFieldDebugMenuPage]
+	dec a
+	cp $ff
+	jr nz, .load_prev_page
+.set_page_2
+	ld a, $02
+.load_prev_page
+	ld [wFieldDebugMenuPage], a
+	jr PlayMenuOpeningSound
+
+FieldDebug_NextFieldDebugMenuPage: ; 3F:4181
+	ld a, [wFieldDebugMenuPage]
+	and a
+	jr z, .set_page_1
+.set_page_0
+	xor a
+	jr .load_page
+.set_page_1
+	ld a, $01
+.load_page
+	ld [wFieldDebugMenuPage], a
+
+PlayMenuOpeningSound: ; 3F:418F
+	ld de, SFX_MENU
+	call PlaySFX
+.loop
+	call GetJoypad
+	ldh a, [hJoyDown]
+	bit 0, a
+	jr nz, .loop
+	ld a, $00
+	ret
+
+CloseFieldDebugMenuTextBoxLoop: ; 3F:41A1
+	push bc
+	ld b, a
+.loop
+	call GetJoypad
+	ld a, [hJoyDown]
+	and b
+	jr z, .loop
+	pop bc
+	ret
+
+ShowTextBoxAndExitFieldDebugMenu: ; 3F:41AD
+	call MenuTextBox
+	ld a, $03
```

Comment:
Just a 3. You get the idea, not going to point them all out. If you know what inmediate values mean and you don't feel creating a constant for them is appropriate don't use the hex notation. Usually the hex notation screams "this code was just pasted from the disassembler in an effort to finish disassembling quickly", unless, of course, it's appropriate (`and $f` and such come to mind, but even then it might be more appropriate to `maskbits` or use a constant).